include/openenclave/seal.h

```diff
@@ -0,0 +1,299 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+/**
+ * @file
+ *
+ * This file defines constants and structures for sealing APIs.
+ *
+ * Only TEE agnostic definitions should go in this file. SGX specific
+ * definitions should go in sgx/seal.h
+ *
+ */
+#ifndef _OE_SEAL_H
+#define _OE_SEAL_H
+
+#include <openenclave/bits/evidence.h>
+#include <openenclave/bits/result.h>
+#include <openenclave/bits/types.h>
+
+OE_EXTERNC_BEGIN
+
+/**
+ * Seal settings as TLV tuples.
+ *
+ * It's strongly recommended to use OE_SEAL_SET_* helper macros to set up this
+ * structure, rather than assigning to members directly.
+ */
+typedef struct _oe_seal_setting
+{
+    int type;      ///< Setting type. See oe_seal_setting_type_t for details.
+    uint32_t size; ///< Size of the buffer pointed to by \c value.p
+    union {
+        uint64_t q; ///< quad-word value. \c size should be set to \c 0.
+        uint32_t d; ///< double-word value. \c size should be set to \c 0.
+        uint16_t w; ///< word value. \c size should be set to \c 0.
+        uint8_t b;  ///< byte value. \c size should be set to \c 0.
+        const void*
+            p; ///< buffer. \c size should be set to the buffer size in bytes.
+    } value;
+} oe_seal_setting_t;
+
+/*
+ * TEE agnostic seal settings supported by OE SDK
+ */
+enum oe_seal_setting_type_t
+{
+    /**
+     * Desired sealing policy - either \c OE_SEAL_POLICY_UNIQUE or \c
+     * OE_SEAL_POLICY_PRODUCT.
+     */
+    OE_SEAL_SETTING_POLICY,
+
+    /**
+     * Additional context to be included in seal key derivation. Please note
+     * that not every seal plug-in supports this setting, in which case
+     * \coe_seal() will return \c OE_UNSUPPORTED. It's highly recommended NOT
+     * to specify this setting.
+     */
+    OE_SEAL_SETTING_ADDITIONAL_CONTEXT,
+
+    /**
+     * IV (Initial Vector) to use should the underlying cipher requires one.
+     * It's highly recommended NOT to specify this setting.
+     */
+    OE_SEAL_SETTING_IV,
+
+    /**
+     * Upper bound of seal setting types.
+     *
+     * No seal setting types shall be defined at or above this.
+     */
+    OE_SEAL_SETTING_MAX = 0x10000,
```

Comment:
I think the feedback from @dthaler based on the last SIG-arch meeting was that these should not be part of the enum definition, but should be standalone constants.

Reply:
The original comment was based on MAX being the current number of entries in the enum.  Now that it's a high number that will never change, it's not as big a deal.  E.g., hopefully no one will consider using it to size an array.  Still: OE_SEAL_SETTING_PLUGIN_DEFINED i understand, but why is OE_SEAL_SETTING_MAX needed at all? (as opposed to just using the natural maxint as the upper bound)

Reply:
That was from @CodeMonkeyLeet's comment in last week's sig-arch meeting. 64K was the suggestion as captured in the meeting notes.